kythe/go/serving/pipeline/beam.go

```diff
@@ -503,3 +513,118 @@ func refKind(r *ppb.Reference) string {
 	}
 	return r.GetGenericKind()
 }
+
+// Edges returns a Kythe edges table derived from the Kythe input graph.  The beam.PCollections have
+// elements of type KV<string, *srvpb.PagedEdgeSet> and KV<string, *srvpb.EdgePage>, respectively.
+func (k *KytheBeam) Edges() (beam.PCollection, beam.PCollection) {
+	s := k.s.Scope("Edges")
+
+	nodes := beam.ParDo(s, moveSourceToKey, k.nodes)
+	edges := beam.ParDo(s, reverseEdge, beam.CoGroupByKey(s, nodes, beam.ParDo(s, nodeToEdges, k.nodes)))
+	rev := beam.ParDo(s, nodeToReverseEdges, k.nodes)
+
+	return beam.ParDo2(s, groupEdges, beam.CoGroupByKey(s, nodes, edges, rev))
+}
+
+// nodeToReverseEdges emits an *ppb.Edge with its SourceNode populated for each of n's edges.  The
+// key for each *ppb.Edge is its Target VName.
+func nodeToReverseEdges(n *ppb.Node, emit func(*spb.VName, *ppb.Edge)) {
+	for _, e := range n.Edge {
+		emit(e.Target, &ppb.Edge{
+			SourceNode: n,
+			Target:     e.Target,
+			Kind:       e.Kind,
+			Ordinal:    e.Ordinal,
+		})
+	}
+}
+
+// nodeToEdges emits an *ppb.Edge for each of n's edges.  The key for each *ppb.Edge is its Target
+// VName.
+func nodeToEdges(n *ppb.Node, emit func(*spb.VName, *ppb.Edge)) {
+	for _, e := range n.Edge {
+		emit(e.Target, &ppb.Edge{
+			Source:  n.Source,
+			Target:  e.Target,
+			Kind:    e.Kind,
+			Ordinal: e.Ordinal,
+		})
+	}
+}
+
+// reverseEdge emits the reverse of each *ppb.Edge, embedding the associated TargetNode.
+func reverseEdge(src *spb.VName, nodeStream func(**ppb.Node) bool, edgeStream func(**ppb.Edge) bool, emit func(*spb.VName, *ppb.Edge)) {
+	var node *ppb.Node
+	for nodeStream(&node) {
```

Comment:
could you express this as follows?

    if !nodeStream(&node) {
      node = &ppb.Node{Source: src}
    }

That seems a bit more concise.

Reply:
✓